### ISSUE TITLE
ErrorBoundary support for multiple exceptions

### ISSUE THREAD

**Issue by thirstyape** [dotnet/aspnetcore#38905]
Am I using this incorrectly? I implemented as per [Documentation](https://docs.microsoft.com/en-us/aspnet/core/blazor/fundamentals/handle-errors?view=aspnetcore-6.0#error-boundaries) in `MainLayout.razor` with the slight change that in my `<ErrorContent>` section I added a button that triggers `errorBoundary?.Recover();`. 

My test involved throwing a generic exception with a message from a child component in `protected override void OnInitialized()`. The result was that instead of displaying the content of `@Body` or the ErrorContent, there was nothing. Upon opening the browser inspector I had an empty `<main></main>`.

I also tried creating a custom class that implemented ErrorBoundary, but this did not help either.

_Originally posted by @thirstyape in https://github.com/dotnet/aspnetcore/issues/30874#issuecomment-989263259_

**Comment by TanayParikh** [comment]
Does it work as expected if you throw the exception in `OnParametersSet` instead of `OnInitialized`? That's the event used in the docs and occurs [__after__](https://docs.microsoft.com/en-us/aspnet/core/blazor/components/lifecycle?view=aspnetcore-6.0#lifecycle-events) `OnInitialized`, so I wonder if this is an event ordering issue.

**Comment by thirstyape** [comment]
Yes, it occurred in `OnParametersSet` as well. That was actually the first time I noticed this. I tested a basic scenario on `Index.razor` in its `OnIntialized` event and that worked as expected. Then I changed a different component from using `OnIntialized` to `OnParametersSet` which threw an exception (it should have). This caused the issue where there was nothing in the `<main>`. I then modified the event code to throw a generic exception instead of its normal actions, which caused the issue regardless of which event I used.

For reference, the component that is seeing issue is a Tab within a TabBar. A merge of the component tree has code similar to the following, I can provide more detailed samples if anything could use further investigation:

```
<MainLayout>
  <ErrorBoundary>
    <ChildContent>
       <UserPage>
          <TabBar>
            <Tab>
            </Tab>
            <Tab>
            </Tab>
          </TabBar>
       </UserPage>
    </ChildContent>
    <ErrorContent>
       <button @onclick="Reset">Reset Error</button>
    </ErrorContent>
  <ErrorBoundary>
</MainLayout>
```

The `TabBar` has `<CascadingValue Value="this">` that contains the ChildContent for the `Tab` items. The `Tab` items add themselves to a collection in the `TabBar` in the event call.

**Comment by TanayParikh** [comment]
Could you please provide a **minimal** (using the base Blazor template with minimal additions) public github repro project to allow us to investigate further?

**Comment by thirstyape** [comment]
Done, here is [the repository](https://github.com/thirstyape/ErrorBoundary-Test).

**Changes to default template:**

- Add TabBar.razor and TabBar.razor.cs
- Add Tab.razor
- Update MainLayout.razor to include ErrorBoundary
- Update Index.razor to include TabBar
- Remove default error notice from index.html

**Steps to reproduce:**

1. Go into Tab.razor and change the `OnInitialized` to `OnParametersSet`
2. Run the app
3. Click the "Click me" text

This will produce some exceptions and the `<article>` will be empty. Interestingly, navigating to another component will correct the behavior and the Failed message appears.

**Unexpected behavior:**
I am not sure why, but when I started the app the `Tab` items did not appear. I needed to add the "Click me" tab directly in `TabBar.razor` and then click it to get the rest of the tabs to appear. After that they work as expected.

This is making me think the `TabBar` or `Tab` components may not be structured correctly.

**Comment by TanayParikh** [comment]
Thanks @thirstyape for the repro. I took a look and I believe it _may_ be because in your example multiple exceptions are being thrown. If we update your example to just throw a single exception, things work as expected (`Failed` is shown).

```diff
diff --git a/BlazorApp1/Shared/Tab.razor b/BlazorApp1/Shared/Tab.razor
index f59c9e1..475e711 100644
--- a/BlazorApp1/Shared/Tab.razor
+++ b/BlazorApp1/Shared/Tab.razor
@@ -16,8 +16,8 @@
     [CascadingParameter]
     private TabBar Parent { get; set; }
 
-    protected override void OnInitialized()
-    //protected override void OnParametersSet() // Changing to this will cause the issue
+    // protected override void OnInitialized()^M
+    protected override void OnParametersSet() // Changing to this will cause the issue^M
     {
         Parent.Tabs.Add(Name, Icon);
     }
diff --git a/BlazorApp1/Shared/TabBar.razor.cs b/BlazorApp1/Shared/TabBar.razor.cs
index 24a6221..576ae92 100644
--- a/BlazorApp1/Shared/TabBar.razor.cs
+++ b/BlazorApp1/Shared/TabBar.razor.cs
@@ -39,8 +39,7 @@ public partial class TabBar
 
     private void OnTabBarChanged(string tab)
     {
-        Active = tab;
-        StateHasChanged();
+               throw new Exception("Yay for exceptions");^M
     }
```

> This is making me think the TabBar or Tab components may not be structured correctly.

I believe this may be the case as well, I skimmed your code, but didn't immediately see what you were using to trigger the exception. My interpretation is the `OnParametersSet` is being triggered multiple times and that's causing a duplicate key to be added to the `Parent.Tabs` leading to the (multiple) exceptions. I'd recommend reviewing your code in that case. If you still believe this to be a framework bug, we'll need an even simpler repro app that doesn't require understanding the heuristics of TabBar, Tab, etc. 

> _may_ be because in your example multiple exceptions are being thrown

@SteveSandersonMS are there any restrictions with respect to ErrorBoundary's and multiple exceptions?

**Comment by TanayParikh** [comment]
I'll leave this open to confirm if there are any restrictions with respect to ErrorBoundary's and multiple exceptions?

**Comment by thirstyape** [comment]
@TanayParikh I did some further testing today and I believe that we can close this out. I was able to use the sample project and get everything to work as expected as long as there was a single exception.

I pushed my changes to the test. With these updates, the `ErrorBoundary` now works as expected and can be re-used as many times as needed. That is, it will not stop working after the first time. Simply clicking the reset button I added will allow the process to repeat.

I think we can reasonably conclude that the multiple exceptions was the cause and is not currently a supported scenario.

**Comment by TanayParikh** [comment]
Thanks @thirstyape, let's keep this issue open to track potentially adding support for multiple exceptions in the future.

**Comment by TanayParikh** [comment]
@SteveSandersonMS would you like to keep this open?

**Comment by SteveSandersonMS** [comment]
Sure, let's reopen and keep on backlog.

**Comment by eduard-tapkey** [comment]
Are there any updates on this topic? One year later it seems multiple exceptions at once are not supported by ErrorBoundary. The ErrorContent won't render.

**Comment by zerox981** [comment]
This bug is making the ErrorBoundary useless for real world scenarios. As soon as we have more than 1 component failing (in OnInitialized(Async), OnParameterSet(Async), nothing gets rendered on the screen (the render code is executed, but there is no output).

**Comment by guardrex** [comment]
I'm tagging myself here to keep an 👁️ on this per discussion at https://github.com/dotnet/aspnetcore/issues/39814#issuecomment-3480951148 indicating that at least one Error Boundary scenario remains to be addressed, probably tracked by this issue.
